book/working_with_tables.md

```diff
@@ -390,3 +394,191 @@ You can also [`reject`](/commands/docs/reject.md) columns in a table by passing
+
+       ```nu
+       ls | upsert 3.index { "--->" } | first 5
+       # => ╭──────┬─────────────────┬──────┬─────────┬──────────────╮
+       # => │    # │      name       │ type │  size   │   modified   │
+       # => ├──────┼─────────────────┼──────┼─────────┼──────────────┤
+       # => │    0 │ CNAME           │ file │    15 B │ 9 months ago │
+       # => │    1 │ CONTRIBUTING.md │ file │ 4.3 KiB │ 9 hours ago  │
+       # => │    2 │ LICENSE         │ file │ 1.0 KiB │ 9 months ago │
+       # => │ ---> │ README.md       │ file │ 2.2 KiB │ 3 weeks ago  │
+       # => │    4 │ assets          │ dir  │ 4.0 KiB │ 9 months ago │
+       # => ╰──────┴─────────────────┴──────┴─────────┴──────────────╯
+
+       ls | upsert 3.index { "--->" } | first 5 | describe
+       # => list<any> (stream)
+
+       ls | upsert 3.index { "--->" } | select index name
+       # Error: cannot find column 'index'
+
+       ls | upsert 3.index { "--->" } | select index? name | first 5
+       # => ╭──────┬─────────────────╮
+       # => │    # │      name       │
+       # => ├──────┼─────────────────┤
+       # => │      │ CNAME           │
+       # => │      │ CONTRIBUTING.md │
+       # => │      │ LICENSE         │
+       # => │ ---> │ README.md       │
+       # => │      │ assets          │
+       # => ╰──────┴─────────────────╯
+       ```
+
+   - As demonstrated in the example above, any rows (records) in the table without an `index` key will continue to display the internal representation.
+
+#### Additional Index Examples
+
+##### Convert # to Index
+
+A useful pattern for converting an internal `#` into an index for all rows, while maintaining the original numbering, is:
+
+```nu
+ls | enumerate | flatten
+```
+
+While the results _look_ the same, the `index` is now decoupled from the internal cell-path. For example:
+
+```nu
+ls | enumerate | flatten | sort-by modified | first 5
+# => ╭────┬──────────────┬──────┬─────────┬──────────────╮
+# => │  # │     name     │ type │  size   │   modified   │
+# => ├────┼──────────────┼──────┼─────────┼──────────────┤
+# => │  0 │ CNAME        │ file │    15 B │ 9 months ago │
+# => │  2 │ LICENSE      │ file │ 1.0 KiB │ 9 months ago │
+# => │  4 │ assets       │ dir  │ 4.0 KiB │ 9 months ago │
+# => │ 17 │ lefthook.yml │ file │ 1.1 KiB │ 9 months ago │
+# => │ 24 │ snippets     │ dir  │ 4.0 KiB │ 9 months ago │
+# => ╰────┴──────────────┴──────┴─────────┴──────────────╯
+
+ls | enumerate | flatten | sort-by modified | select 4
+# => ╭────┬──────────┬──────┬─────────┬──────────────╮
+# => │  # │   name   │ type │  size   │   modified   │
+# => ├────┼──────────┼──────┼─────────┼──────────────┤
+# => │ 24 │ snippets │ dir  │ 4.0 KiB │ 9 months ago │
+# => ╰────┴──────────┴──────┴─────────┴──────────────╯
+```
+
+The `sort-by modified` now _also_ sorts the `index` along with the rest of the columns.
+
+##### Adding a Row Header
+
+```nu
+let table = [
+[additions | deletions | delta ];
```

Comment:
I don't think we should publicize `|` working as a delimiter here. I think it's a bug and probably shouldn't be allowed.

Reply:
You're right, of course.  I've gotten used to it so quickly for table literals that I've almost forgotten it's not supported.  It's definitely a bug, as the parse code specifically *doesn't* include `'|` as a list separator today.  But I'm going to propose that we just fix the bug by changing the parser to actually allow it.  I already have a fix in and tested, so I'll drop a PR soon and see if it flies.

In the meantime, I agree that we shouldn't document it until and unless it's official, so I'll reformat this.